usr/src/uts/common/fs/zfs/dbuf.c

```diff
@@ -918,7 +920,7 @@ dbuf_free_range(dnode_t *dn, uint64_t start_blkid, uint64_t end_blkid,
 	db_search.db_state = DB_SEARCH;
 
 	mutex_enter(&dn->dn_dbufs_mtx);
-	if (start_blkid >= dn->dn_unlisted_l0_blkid) {
+	if (start_blkid >= dn->dn_unlisted_l0_blkid && !freespill) {
```

Comment:
This seems fine.  FYI, we are going to remove dn_unlisted_l0_blkid, having measured that its performance improvement is no longer required now that dn_dbufs is an avl tree.